esp-hal/src/dma/buffers.rs

```diff
@@ -451,15 +651,22 @@ impl DmaRxBuf {
             .sum::<usize>()
     }
 
+    fn set_length_fallible(&mut self, len: usize, burst: BurstConfig) -> Result<(), DmaBufError> {
+        assert!(len <= self.buffer.len());
+
+        self.descriptors.set_rx_length(
+            len,
+            burst.max_chunk_size_for(self.buffer, TransferDirection::In),
+        )
+    }
```

Comment:
Please clarify which device needs exact block lengths, in which circumstances, and where the TRMs specify this.

Reply:
All of them. For example, in ESP32-S3 TRM, section 3.4.8, table 3-2.

When doing burst transfer in RX mode, the buffer address pointer and size must be word aligned.

`set_rx_length` just splits the `len` into chunks. Since the chunk sizes are aligned, the buffer pointers will be correctly aligned. However the very last chunk could be less than the chunk size, and as such won't be aligned correctly.

Take this real example. Say we have a `len` of 8186. The "max chunk size" for RX burst is 4092, which is word-aligned.
The `len` of 8186 will be split into chunks of 4092. Which will produce three chunks. 4092, 4092 and 2. The third chunk/descriptor of size 2 is not word-aligned, which is not allowed.

Reply:
Please don't mix up size and length. There is no restriction on length, and size must be aligned so that the DMA doesn't accidentally write outside of the given buffer.

Reply:
Ehhh, the names are confusing at this level of `set_length_fallible` but I promise I'm referring to descriptor size and not the descriptor length.

To help clear things up quickly, remember that when doing an RX transfer, you always set the length of each descriptor to 0 and the DMA will actually populate it with how many bytes it wrote to the descriptor. (See `DmaDescriptor::reset_for_rx`).
For RX you can only control size, not length. It's TX that you can control length and size.

So when we're talking about the "length" up here in `set_length_fallible`, we're actually talking about the descriptor sizes.